src/main/scala/leon/comparison/Utils.scala

```diff
@@ -0,0 +1,239 @@
+  def mean(a: Double, b: Double, c: Double): Double = (a + b + c) / 3
+  def mean(a: Double, b: Double, c: Double, d: Double): Double = (a + b + c + d) / 4
+  def mean(list : List[Double]): Double = list.foldLeft(0.0)(_+_) / list.size.toDouble
+
+  /**
+    * Derived from "traverse" function. Traverse all the tree and collect whished information about Expr composing it.
+    * It fixes the "onChildren" function to be recursive and let the "onParent" be the one deciding what information
+    * will be stored
+    *
+    * collectExpr and collectClass collect respectively the Expr and the Class of each element of the tree.
+    *
+    * BEWARE: Expr are complete trees even if we call it "parent". When we compare two Expr, we compare two entire tree.
+    * At the contrary, when we compare to Class, we lose this information and only compare the Class of two parent.
+    * @param expr
+    * @param f
+    * @tparam T
+    * @return
+    */
+  def collect[T](expr: Expr)(f:Expr => List[T]): List[T] = traverse(expr)(f)(expr => collect(expr)(f))
+
+  def collectClass(expr: Expr): List[Class[_ <: Expr]] =
+    collect[Class[_ <: Expr]](expr) (expr => List(expr.getClass))
+
+  def collectExpr(expr: Expr): List[Expr] =
+    collect[Expr](expr) (expr => List(expr))
+
+  /**
+    * Give a list of all children of one parent. Why do we need to use "traverse" function to get them? Because
+    * there is a lot of possible CaseClass extending Expr, and we want to deal with any of them.
+    * @param expr
+    * @return
+    */
+  def getChildren(expr: Expr): List[Expr] =
+    traverse(expr) (expr => Nil) (expr => List(expr))
+
+
+  /**
+    * Main function allowing to traverse a tree. This function is made in a way to be the more generic possible,
+    * letting choose how we deal with the parent (the current Expr) and its children.
+    *
+    * The function "collect" derives from it, by choosing "onChild" to be a recursive call.
+    * The function "getChild" also derives from it, by doing nothing with the parent and adding the children to the
+    * list, but without recursion.
+    *
+    * As we can see, this function allow to do a lot of things and can be used in the future to test new configuration
+    *
+    * @param expr the parent of the tree we want to traverse. Never forget, we use the term "parent" but expr contains
+    *             all the tree
+    * @param onParent function applied to this parent. It can neglect it (expr => Nil), store it (expr => List(expr))
+    *                 or store one of its parameter (expr => List(expr.getClass))
+    * @param onChild function applied to the children expression of "expr". It can be recursive (as in "collect"
+    *                function) or can store some information about it (as in "getChildren)
+    * @tparam T the type of the returned list we want to get after traversing the tree
+    * @return a list containing all the chosen values stored by both function "onParent" and "onChildren"
+    */
```

Comment:
I think this should definitely be implementable with functions in purescala.ExprOps/GenTreeOps